### PR TITLE
[`AutoDocstring`] Based on inspect parsing of the signature

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
- Config stay explicit
- any place that had non explicit (dynamic docstring) gets replaced
- we can leverage inheritance for forward dynamic typing I think

### CODE REVIEW

File: src/transformers/models/align/modeling_align.py [hunk 1]
```diff
@@ -1194,15 +1056,16 @@ def _init_weights(self, module):
...
-    ALIGN_START_DOCSTRING,
-)
+@auto_docstring
 class AlignTextModel(AlignPreTrainedModel):
     config_class = AlignTextConfig
     _no_split_modules = ["AlignTextEmbeddings"]
 
     def __init__(self, config: AlignTextConfig, add_pooling_layer: bool = True):
+        """
+        add_pooling_layer (`bool`, *optional*, defaults to `True`):
```
Comment (ArthurZucker) [inline src/transformers/models/align/modeling_align.py:1066]: perfect!

File: src/transformers/utils/args_doc.py [hunk 2]
```diff
@@ -0,0 +1,1109 @@
...
+
+UNROLL_KWARGS_METHODS = {
+    "preprocess",
+}
+
+UNROLL_KWARGS_CLASSES = {
+    "ImageProcessorFast",
+}
+
+OPTIONAL_STRING = r", *optional*"
```
Comment (ArthurZucker) [inline src/transformers/utils/args_doc.py:43]: these are not super necessary no?

File: src/transformers/utils/args_doc.py [hunk 3]
```diff
@@ -0,0 +1,1109 @@
...
+    Mask to avoid performing attention on padding pixel values. Mask values selected in `[0, 1]`:
+
+    - 1 for pixels that are real (i.e. **not masked**),
+    - 0 for pixels that are padding (i.e. **masked**).
+
+    [What are attention masks?](../glossary#attention-mask)
+    """
+
+
+class ClassDocstring:
```
Comment (ArthurZucker) [inline src/transformers/utils/args_doc.py:375]: maybe an enum would work better here?

File: src/transformers/utils/args_doc.py [hunk 4]
```diff
@@ -0,0 +1,1109 @@
...
+        docstring = ""
+
+    return params, docstring
+
+
+def contains_type(type_hint, target_type) -> Tuple[bool, Optional[object]]:
+    """
+    Check if a "nested" type hint contains a specific target type,
+    return the first-level type containing the target_type if found.
+    """
```
Comment (ArthurZucker) [inline src/transformers/utils/args_doc.py:600]: a small example would be very nice here!

File: src/transformers/utils/args_doc.py [hunk 5]
```diff
@@ -0,0 +1,1109 @@
...
+    file_name = path.split(os.path.sep)[-1]
+    for file_type in AUTODOC_FILES:
+        start = file_type.split("*")[0]
+        end = file_type.split("*")[-1] if "*" in file_type else ""
+        if file_name.startswith(start) and file_name.endswith(end):
+            model_name_lowercase = file_name[len(start) : -len(end)]
+            return model_name_lowercase
+    else:
+        print(f"🚨 Something went wrong trying to find the model name in the path: {path}")
+        return "model"
```
Comment (ArthurZucker) [inline src/transformers/utils/args_doc.py:630]: Let's re-use the one from modular?

File: src/transformers/utils/args_doc.py [hunk 6]
```diff
@@ -0,0 +1,1109 @@
...
+    placeholders_dict = get_placeholders_dict(placeholders, model_name)
+
+    # replace the placeholders in the args with the values from the placeholders_dict
+    for arg in args:
+        new_arg = args[arg]["description"]
+        placeholders = re.findall(r"{(.*?)}", new_arg)
+        placeholders = [placeholder for placeholder in placeholders if placeholder in placeholders_dict]
+        if placeholders:
+            new_arg = new_arg.format(**{placeholder: placeholders_dict[placeholder] for placeholder in placeholders})
+        args[arg]["description"] = new_arg
```
Comment (ArthurZucker) [inline src/transformers/utils/args_doc.py:675]: we should be able to do re.sub here no? (for perfs)!

File: src/transformers/utils/args_doc.py [hunk 7]
```diff
@@ -0,0 +1,1109 @@
...
+
+        </Tip>
+
+        """
+        intro_docstring = equalize_indent(intro_docstring, indent_level + 4)
+
+    return intro_docstring
+
+
+def auto_method_docstring(func, parent_class=None, custom_intro=None, checkpoint=None):
```
Comment (ArthurZucker) [inline src/transformers/utils/args_doc.py:732]: this function is too big! we need to create smaller functions!

File: src/transformers/utils/args_doc.py
Comment (ArthurZucker) [inline src/transformers/utils/args_doc.py:1]: Very nice!

**Review by ArthurZucker — COMMENTED** [review]
Mega nice! just needs: 
- cleanup the big big function to have a func for each part that we are building
- a small benchmark if possible on time it takes to get the doc
- a small doc `.md` to explain how to define new args for example
Super happy about it!

**Review by ArthurZucker — COMMENTED** [review]
SUper nice update! Just missing imo:
- check that we are not compiling, if compiling we don't produce doc probably
- if this is too slow, we can make sure only the `doc-builder` actually generates the doc
🔥 otherwise!

**Comment by yonigozlan** [comment]
> check that we are not compiling, if compiling we don't produce doc probably

Indeed I had some issues with this! When using the `@wraps` decorator in auto_docstring and when compiling, the code would crash with a weird error.
For now I just removed the wraps decorator and it seems to be working.
I'm struggling to catch whether or not we are compiling inside a decorator though for now. 
When testing for `if torch.compiler.is_compiling():` at the very top of the auto_docstring decorator, it's always false, even when running tests like `test_generate_compilation_all_outputs`...

> if this is too slow, we can make sure only the doc-builder actually generates the doc

It's strangely quite fast now! Getting rid of `inspect.getsource` seems to have done the job.
Right now, when profiling `from transformers import *`, I get:
```bash
- total time: 15.581 seconds
- time taken by remaining old docstring decorators: ~2.2 seconds
- time taken by new docstring decorators: ~1.3 seconds
```

On the main branch, I get:
```bash
- total time: 15.581 seconds
- time taken by remaining old docstring decorators: ~3.6 seconds
```

So it might be faster with the new decorators in the end.

**Comment by yonigozlan** [comment]
Hey @stevhliu ! Would you mind having a look at the new `auto_docstring.md` file? Also this PR changes the way the models docstrings are generated so I thought it could be of interest to you!

File: src/transformers/models/bart/modeling_bart.py [hunk 8]
```diff
@@ -1889,14 +1802,33 @@ def forward(
...
+
+            If you want to change padding behavior, you should read [`modeling_bart._prepare_decoder_attention_mask`]
+            and modify to your needs. See diagram 1 in [the paper](https://arxiv.org/abs/1910.13461) for more
+            information on the default strategy.
+        cross_attn_head_mask (`torch.Tensor` of shape `(decoder_layers, decoder_attention_heads)`, *optional*):
+            Mask to nullify selected heads of the cross-attention modules in the decoder. Mask values selected in `[0,
+            1]`:
+
+            - 1 indicates the head is **not masked**,
+            - 0 indicates the head is **masked**.
```
Comment (ArthurZucker) [inline src/transformers/models/bart/modeling_bart.py:1831]: we should register them indeed!

**Review by ArthurZucker — COMMENTED** [review]
:rocket: bug kudos to you! Amazing work!

**Comment by ArthurZucker** [comment]
>So it might be faster with the new decorators in the end. 
Huge!

**Review by LysandreJik — APPROVED** [review]
Very impressive work!

File: src/transformers/models/bart/modeling_bart.py (hunk 8 above)
Comment (yonigozlan) [reply src/transformers/models/bart/modeling_bart.py:1831]: I think they are! but these are overriden as the shape is not the same as in the registered ones (`target_sequence_length` instead of `sequence_length`).
Very small and maybe irrelevant difference here but it was checked by my conversion script 😅. The rule was to override registered argument if the shape is different, or if the description is very different from the registered one

File: src/transformers/models/aya_vision/modular_aya_vision.py [hunk 9]
```diff
@@ -246,20 +139,10 @@ def forward(
...
-                token can save memory, which becomes pretty significant for long sequences or large vocabulary size.
-                If a `torch.Tensor`, must be 1D corresponding to the indices to keep in the sequence length dimension.
-                This is useful when using packed tensor format (single dimension for batch and sequence length).
-
-
-        Returns:
+        labels (`torch.LongTensor` of shape `(batch_size, sequence_length)`, *optional*):
+            Labels for computing the masked language modeling loss. Indices should either be in `[0, ...,
+            config.vocab_size]` or -100 (see `input_ids` docstring). Tokens with indices set to `-100` are ignored
+            (masked), the loss is only computed for the tokens with labels in `[0, ..., config.vocab_size]`.
```
[the remark this answers is not in the document]
Comment (yonigozlan) [reply src/transformers/models/aya_vision/modular_aya_vision.py:142]: Yes but I decided to leave most labels description as they are often different from the registered one

File: src/transformers/models/data2vec/modeling_data2vec_audio.py [hunk 10]
```diff
@@ -42,12 +36,6 @@
 
 logger = logging.get_logger(__name__)
 
-# Base docstring
-_CHECKPOINT_FOR_DOC = "facebook/data2vec-audio-base-960h"
-
-# General docstring
-_CONFIG_FOR_DOC = "Data2VecAudioConfig"
-
```
[the remark this answers is not in the document]
Comment (yonigozlan) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:50]: Yes that could be nice. Right now the checkpoints are parsed from the configs, so we could even generate this file automatically

File: src/transformers/image_processing_utils_fast.py [hunk 11]
```diff
@@ -190,107 +190,7 @@ class DefaultFastImageProcessorKwargs(TypedDict, total=False):
...
-            - `"none"` or `ChannelDimension.NONE`: image in (height, width) format.
-        device (`torch.device`, *optional*, defaults to `self.device`):
-            The device to process the images on. If unset, the device is inferred from the input images."""
-
-
-@add_start_docstrings(
-    "Constructs a fast base image processor.",
-    BASE_IMAGE_PROCESSOR_FAST_DOCSTRING,
-)
+@auto_docstring
```
Comment (pcuenca) [inline src/transformers/image_processing_utils_fast.py:193]: wow amazing

**Comment by sfc-gh-sbekman** [comment]
@ArthurZucker, could there be some additional work done to prevent this:

```
🚨 `image_seq_length` is part of ImagesKwargs, but not documented. Make sure to add it to the docstring of the function in /code/users/stas/github/transformers-alst-integration/src/transformers/image_processing_utils_fast.py.
...
🚨 `image_seq_length` is part of ImagesKwargs, but not documented. Make sure to add it to the docstring of the function in /code/users/stas/github/transformers-alst-integration/src/transformers/image_processing_utils_fast.py.
🚨 `image_seq_length` is part of ImagesKwargs, but not documented. Make sure to add it to the docstring of the function in /code/users/stas/github/transformers-alst-integration/src/transformers/image_processing_utils_fast.py.
```
one for each gpu.

I'm not using  `image_seq_length` or dealing with images, just doing a post-training of `Qwen/Qwen3-30B-A3B-Instruct-2507` with text.

1. This warning is not actionable for the end user. If it should be actionable by users it should show the trace to pinpoint the offensive caller.
3. If this message is for maintainers of some packages, why is it being shown to end users?
5. Bonus for removing `🚨` - there is enough pollution in the logs multiplied by many gpus to deal with attention pulling unicode. Please don't set a bad precendent for other package managers to copy, we users are drowning in a sea of random messages many packages emit multiplied by the number of gpus.

Apologies if this comment feels frustrated, I'm not sure what else to do other than responding to each of these events as they happen and asking the maintainers to please think about the users when deciding to emit warnings.

Thank you.

**Comment by yonigozlan** [comment]
Sorry about this, this will be fixed in https://github.com/huggingface/transformers/pull/42054
As this warning is meant for contributor, it will be changed to an error in this PR https://github.com/huggingface/transformers/pull/41635, so contributors won't be able to merge without addressing an undocumented arg issue. 
Hopefully this means you won't see this warning again!

**Comment by stas00** [comment]
super appreciating this, @yonigozlan! Thank you for fixing both issues!
